### PR TITLE
Fixes #1363. The main change to create a copy of the exports, so we c…

### LINKED ISSUE #1363: Change naming convention for split fields in HISTORY
The current implementation of generating names for the split fields cannot produce in certain situations the desired outcome (i.e. preserve the file spec). The main issue is that if it is desired to generate a split name is identical to the original un-split, this would create a conflict due to duplicates.

### PR DESCRIPTION
**Pull request by atrayano**
…ould avoid duplicates. Also change the strategy for name generation

## Description

Created a copy of the export state(s) to avoid duplicate field name(s) . Changed naming convention to not  alter field alias

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Trivial change (affects only documentation or cleanup)

## Checklist:

- [x] I have tested this change with a run of GEOSgcm (if non-trivial)
- [x] I have added one of the required labels (0 diff, 0 diff trivial, 0 diff structural, non 0-diff)
- [x] I have updated the CHANGELOG.md accordingly following the style of [Keep a Changelog](https://keepachangelog.com/en/1.0.0/#how)

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -23,6 +23,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Changed
 
+- Now History creates a copy of the export state(s) so that we can remove the original un-split field to avoid duplicates
+- Changed the naming convention for the split name(s)
```
Comment (tclune) [inline CHANGELOG.md:27]: This one is better, but probably worth stating _how_ that convention changed.

**Review by tclune — DISMISSED** [review]

File: CHANGELOG.md [hunk 2]
```diff
@@ -23,6 +23,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Changed
 
+- Now History creates a copy of the export state(s) so that we can remove the original un-split field to avoid duplicates
```
[the remark this answers is not in the document]
Comment (atrayano) [reply CHANGELOG.md:26]: This is fixing the case when the user wants one of the split names to be the same as the original. I'll try to reword the CHANGELOG

**Comment by mathomp4** [comment]
I am checking this now with GOCART2G as well as regular GEOS.

**Comment by mathomp4** [comment]
Putting a blocker on this. GOCART2G was not happy:
```
DU::Run1                                       192
pe=00060 FAIL at line=00075    GetPointer.H                             <status=13>
pe=00060 FAIL at line=00653    DU2G_GridCompMod.F90                     <status=13>
pe=00060 FAIL at line=01799    MAPL_Generic.F90                         <status=13>
 Integer*4 Resource Parameter: ORB2B_REF_YYYYMMDD:20000101
DU::Run1                                       192
...
pe=00056 FAIL at line=00075    GetPointer.H                             <status=13>
pe=00056 FAIL at line=00602    SS2G_GridCompMod.F90                     <status=13>
pe=00056 FAIL at line=01799    MAPL_Generic.F90                         <status=13>
SS::Run1                                        74
...
pe=00092 FAIL at line=00167    BaseProfiler.F90                         <Timer <GOCART2G> does not match start timer <NI>>
...
NI::Run1                                       100
PHYSICS::Run                                  2575
AGCMRun                                       2382
GCMRun                                        1803
...
```

I did re-test with MAPL `develop` and it did run, so it's something in this PR.

**Comment by atrayano** [comment]
@tclune I am done with the revised gocart split code. The new code has some ugliness - I had to grow few arrays (r8_to_r4 and similar) but I did not want to do the split multiple time just to get an accurate count of the size of the new field_set (this gets adjusted, but it has to be done later, since we were inside a loop, and the loop bunds were using the field_set.
@mathomp4 I just pushed my gocart related changes. Could you check it again with your reproducer?

**Review by tclune — APPROVED** [review]
Ugh.

**Comment by mathomp4** [comment]
> @tclune I am done with the revised gocart split code. The new code has some ugliness - I had to grow few arrays (r8_to_r4 and similar) but I did not want to do the split multiple time just to get an accurate count of the size of the new field_set (this gets adjusted, but it has to be done later, since we were inside a loop, and the loop bunds were using the field_set. @mathomp4 I just pushed my gocart related changes. Could you check it again with your reproducer?

@atrayano I'll test it soon (might be Monday). Working with @sdrabenh to get GOCART2G onto GEOSgcm `main`. We'll wait until we get a good MAPL 2.18.0 before we finally release GOCART2G GEOS

**Comment by mathomp4** [comment]
Well, good news, bad news, @atrayano. This branch does not crash, but I'm not sure it's doing what we want.

Looking at the tavg2d history output, most fields seem to be in both, but not all:
```
$ sdiff -w 40 ~/stock_fields.dat ~/atanas_fields.dat | grep '|'
BCEXTTAU001	   |	BCEXTTAU
BCSCATAU001	   |	BCSCATAU
BREXTTAU001	   |	BREXTTAU
BRSCATAU001	   |	BRSCATAU
DUDP001		   |	DUDP
DUEM001		   |	DUEM
DUEXTT25001	   |	DUEXTT25
DUEXTTAU001	   |	DUEXTTAU
DUEXTTFM001	   |	DUEXTTFM
DUSCAT25001	   |	DUSCAT25
DUSCATAU001	   |	DUSCATAU
DUSCATFM001	   |	DUSCATFM
DUSD001		   |	DUSD
DUWT001		   |	DUWT
NIDP001		   |	NIDP
NIEXTTAU001	   |	NIEXTTAU
NIEXTTFM001	   |	NIEXTTFM
NIHT001		   |	NIHT
NISCATAU001	   |	NISCATAU
NISCATFM001	   |	NISCATFM
NISD001		   |	NISD
NIWT001		   |	NIWT
OCEXTTAU001	   |	OCEXTTAU
OCSCATAU001	   |	OCSCATAU
SSDP001		   |	SSDP
SSEM001		   |	SSEM
SSEXTT25001	   |	SSEXTT25
SSEXTTAU001	   |	SSEXTTAU
SSEXTTFM001	   |	SSEXTTFM
SSSCAT25001	   |	SSSCAT25
SSSCATAU001	   |	SSSCATAU
SSSCATFM001	   |	SSSCATFM
SSSD001		   |	SSSD
SSWT001		   |	SSWT
SUDP001		   |	SUDP
SUEM001		   |	SUEM
SUEXTTAU001	   |	SUEXTTAU
SUSCATAU001	   |	SUSCATAU
SUWT001		   |	SUWT
TOTEXTTAU001	   |	TOTEXTTAU
TOTSCATAU001	   |	TOTSCATAU
```
So something like:
```
SUEXTTAU001	   |	SUEXTTAU
```
seems right. But then in some cases:
```
SUEM001		   |	SUEM
SUEM002			SUEM002
SUEM003			SUEM003
SUEM004			SUEM004
```
So it's like all `001` fields are now no longer `001` and I'm not sure that's what is wanted.

**Comment by mathomp4** [comment]
Also, in the section of the log where it says:
```
 FV3 is Advecting the following    48 tracers:
```
Again, most is the same, but I see two differences:
```
$ sdiff ~/stock_advect.txt ~/atanas_advect.txt -w 80 | grep '|'
 DU::DU001			      |	 Field4034
 SS::SS001			      |	 Field4046
```
The left column is MAPL 2.17.2, the right is this branch.

**Comment by atrayano** [comment]
Matt, the new behavior is intentional – Arlindo wanted to be able to “match” the old file spec exactly, i.e. if alias is provided, I take it as is, and I am not appending digits (unless the user did not specify enough alias fields)


From: Matthew Thompson ***@***.***>
Reply-To: GEOS-ESM/MAPL ***@***.***>
Date: Monday, February 21, 2022 at 9:11 AM
To: GEOS-ESM/MAPL ***@***.***>
Cc: "Trayanov, Atanas L. (GSFC-610.1)[SCIENCE SYSTEMS AND APPLICATIONS INC]" ***@***.***>, Mention ***@***.***>
Subject: [EXTERNAL] Re: [GEOS-ESM/MAPL] Fixes #1363. The main change to create a copy of the exports, so we c… (PR #1364)


Well, good news, bad news, @atrayano<https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fatrayano&data=04%7C01%7CAtanas.L.Trayanov%40nasa.gov%7C22467b5ff7734ec3c44808d9f5441b0e%7C7005d45845be48ae8140d43da96dd17b%7C0%7C0%7C637810495138114473%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=lbIeAjwftKSBvs0bxplz%2BQumakBeH9EZ1xp3E%2BP4kpg%3D&reserved=0>. This branch does not crash, but I'm not sure it's doing what we want.

Looking at the tavg2d history output, most fields seem to be in both, but not all:

$ sdiff -w 40 ~/stock_fields.dat ~/atanas_fields.dat | grep '|'

BCEXTTAU001       |    BCEXTTAU

BCSCATAU001       |    BCSCATAU

BREXTTAU001       |    BREXTTAU

BRSCATAU001       |    BRSCATAU

DUDP001           |    DUDP

DUEM001           |    DUEM

DUEXTT25001       |    DUEXTT25

DUEXTTAU001       |    DUEXTTAU

DUEXTTFM001       |    DUEXTTFM

DUSCAT25001       |    DUSCAT25

DUSCATAU001       |    DUSCATAU

DUSCATFM001       |    DUSCATFM

DUSD001           |    DUSD

DUWT001           |    DUWT

NIDP001           |    NIDP

NIEXTTAU001       |    NIEXTTAU

NIEXTTFM001       |    NIEXTTFM

NIHT001           |    NIHT

NISCATAU001       |    NISCATAU

NISCATFM001       |    NISCATFM

NISD001           |    NISD

NIWT001           |    NIWT

OCEXTTAU001       |    OCEXTTAU

OCSCATAU001       |    OCSCATAU

SSDP001           |    SSDP

SSEM001           |    SSEM

SSEXTT25001       |    SSEXTT25

SSEXTTAU001       |    SSEXTTAU

SSEXTTFM001       |    SSEXTTFM

SSSCAT25001       |    SSSCAT25

SSSCATAU001       |    SSSCATAU

SSSCATFM001       |    SSSCATFM

SSSD001           |    SSSD

SSWT001           |    SSWT

SUDP001           |    SUDP

SUEM001           |    SUEM

SUEXTTAU001       |    SUEXTTAU

SUSCATAU001       |    SUSCATAU

SUWT001           |    SUWT

TOTEXTTAU001      |    TOTEXTTAU

TOTSCATAU001      |    TOTSCATAU

So something like:

SUEXTTAU001       |    SUEXTTAU

seems right. But then in some cases:

SUEM001           |    SUEM

SUEM002                SUEM002

SUEM003                SUEM003

SUEM004                SUEM004

So it's like all 001 fields are now no longer 001 and I'm not sure that's what is wanted.

—
Reply to this email directly, view it on GitHub<https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FGEOS-ESM%2FMAPL%2Fpull%2F1364%23issuecomment-1046921177&data=04%7C01%7CAtanas.L.Trayanov%40nasa.gov%7C22467b5ff7734ec3c44808d9f5441b0e%7C7005d45845be48ae8140d43da96dd17b%7C0%7C0%7C637810495138114473%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=yKlPRs748SYKB2T0qCn83OQKdM3Gv%2FOCmxbchKqgxrs%3D&reserved=0>, or unsubscribe<https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAL6ZCVNE3SZQOVGQ75E4QJDU4JB2JANCNFSM5OPK5ANA&data=04%7C01%7CAtanas.L.Trayanov%40nasa.gov%7C22467b5ff7734ec3c44808d9f5441b0e%7C7005d45845be48ae8140d43da96dd17b%7C0%7C0%7C637810495138114473%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=FeOnxIe66UL4i%2Fg4dma2JxtkXUdDcY6xsV4aDq3%2F460%3D&reserved=0>.
Triage notifications on the go with GitHub Mobile for iOS<https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fapps.apple.com%2Fapp%2Fapple-store%2Fid1477376905%3Fct%3Dnotification-email%26mt%3D8%26pt%3D524675&data=04%7C01%7CAtanas.L.Trayanov%40nasa.gov%7C22467b5ff7734ec3c44808d9f5441b0e%7C7005d45845be48ae8140d43da96dd17b%7C0%7C0%7C637810495138114473%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=BBgswCmqP9OaVoW3a1B7emEEiaQ2RAZBXfezjdV93U8%3D&reserved=0> or Android<https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fplay.google.com%2Fstore%2Fapps%2Fdetails%3Fid%3Dcom.github.android%26referrer%3Dutm_campaign%253Dnotification-email%2526utm_medium%253Demail%2526utm_source%253Dgithub&data=04%7C01%7CAtanas.L.Trayanov%40nasa.gov%7C22467b5ff7734ec3c44808d9f5441b0e%7C7005d45845be48ae8140d43da96dd17b%7C0%7C0%7C637810495138114473%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=1k8n%2BmJR4JlB%2FOgI3fR4CG3M0rZL2BlfFiPvmZ4TCyU%3D&reserved=0>.

**Comment by atrayano** [comment]
Matt, the new behavior is intentional – Arlindo wanted to be able to “match” the old file spec exactly, i.e. if alias is provided, I take it as is, and I am not appending digits (unless the user did not specify enough alias fields)

**Comment by atrayano** [comment]
@mathomp4 Just pushed a new version - handling the case the aliasName is not passed in to MAPL_FieldSplit. Hopefully this will take care of the "Fieldnnnn" bug

**Comment by mathomp4** [comment]
> @mathomp4 Just pushed a new version - handling the case the aliasName is not passed in to MAPL_FieldSplit. Hopefully this will take care of the "Fieldnnnn" bug

Okay. I'll test soon!

**Comment by mathomp4** [comment]
Looks good with Intel. I'm building now with GNU to make sure that works as well.

**Comment by atrayano** [comment]
I hope it works with GNU as well. Thanks for testing!

**Comment by mathomp4** [comment]
> I hope it works with GNU as well. Thanks for testing!

GNU seemed quite unhappy. I'm building now with Debug flags so I can get more info.

**Comment by mathomp4** [comment]
More info from GNU:
```
Error termination. Backtrace:
At line 3429 of file /discover/swdev/mathomp4/Models/GEOSgcm-AtanasBranch2810-GNU/GEOSgcm/src/Shared/@MAPL/gridcomps/History/MAPL_HistoryGridComp.F90
Fortran runtime error: Index '150' of dimension 1 of array 'list%r8_to_r4' above upper bound of 149

#0  0x2ba7b21e61dc in run
	at /discover/swdev/mathomp4/Models/GEOSgcm-AtanasBranch2810-GNU/GEOSgcm/src/Shared/@MAPL/gridcomps/History/MAPL_HistoryGridComp.F90:3429
#1  0x2ba7b6b0b1f9 in ???
#2  0x2ba7b6b0b56d in ???
#3  0x2ba37742c20e in ???
```

So...yeah. Definitely @atrayano level.

I'm going to run your branch with Intel Debug as well and see what it says. (Shame on me for using Release first!)

**Comment by mathomp4** [comment]
From Intel:
```
forrtl: severe (408): fort: (2): Subscript #1 of the array R8_TO_R4 has value 150 which is greater than the upper bound of 149

Image              PC                Routine            Line        Source             
GEOSgcm.x          000000000F3BB82F  Unknown               Unknown  Unknown
libMAPL.history.s  00002B49689DF61B  mapl_historygridc        3429  MAPL_HistoryGridComp.F90
libesmf.so         00002B497049E514  _ZN5ESMCI6FTable1     Unknown  Unknown
libesmf.so         00002B49704A205F  ESMCI_FTableCallE     Unknown  Unknown
libesmf.so         00002B4970A3DA8A  _ZN5ESMCI3VMK5ent     Unknown  Unknown
libesmf.so         00002B4970A2B285  _ZN5ESMCI2VM5ente     Unknown  Unknown
libesmf.so         00002B497049F96A  c_esmc_ftablecall     Unknown  Unknown
libesmf.so         00002B4970CD286D  esmf_compmod_mp_e     Unknown  Unknown
libesmf.so         00002B4970FE1C51  esmf_gridcompmod_     Unknown  Unknown
libMAPL.generic.s  00002B49698096F4  mapl_genericmod_m        1794  MAPL_Generic.F90
```

**Comment by atrayano** [comment]
@mathomp4  thanks for for the traceback! I'll take a look

**Comment by atrayano** [comment]
@mathomp4 Fixed my logic. Ready for your tests

**Pull request merged**